packages/@lwc/engine-dom/src/custom-elements/create-custom-element.ts

```diff
@@ -96,13 +105,14 @@ export const createCustomElement = (
 ) => {
     const UpgradableConstructor = getUpgradableConstructor(tagName, isFormAssociated);
 
+    if (Boolean(UpgradableConstructor.formAssociated) !== isFormAssociated) {
+        throw new Error(
+            `<${tagName}> was already registered with formAssociated=${UpgradableConstructor.formAssociated}. It cannot be re-registered with formAssociated=${isFormAssociated}. Please rename your component to have a different name than <${tagName}>`
+        );
+    }
+
```

Comment:
This is not a perf optimization – I just felt it was odd to have this logic inside of the `try`/`finally`, since it doesn't strictly need to be there. :)

Note that the `UpgradableConstructor.formAssociated` now has to be cast to a `Boolean` because it may be `undefined`.